File: 정미광/PGS/게임맵최단거리.java
import java.util.*;

class Solution {
    public int solution(int[][] maps) {
        final int[] moved_x = {-1, 1, 0, 0};
        final int[] moved_y = {0, 0, -1, 1};
        int answer = -1;
        int n = maps.length;
        int m = maps[0].length;
        boolean[][] visited = new boolean[n][m];
        Queue<int[]> next = new LinkedList<>();

        next.add(new int[]{0, 0, 1});
        visited[0][0] = true;
        while (!next.isEmpty()) {
            int[] current = next.poll();
            int x = current[0];
            int y = current[1];
            int moved = current[2];

            if (x == n - 1 && y == m - 1) {
                return moved;
            }

            for (int i = 0; i < 4; i++) {
                int dx = x + moved_x[i];
                int dy = y + moved_y[i];

                if (isInRange(dx, dy, n, m) && !visited[dx][dy] && maps[dx][dy] == 1) {
                    visited[dx][dy] = true;
                    next.add(new int[]{dx, dy, moved + 1});
                }
            }
        }

        return answer;
    }

    boolean isInRange(int x, int y, int n, int m) {
        if (x < 0 || x >= n) {
            return false;
        }
        if (y < 0 || y >= m) {
            return false;
        }
        return true;
    }
}

File: 권나연/BFSDFS/게임맵최단거리.py
from collections import deque

dx = [0, 0, -1, 1]
dy = [1, -1, 0, 0]

def bfs(maps):
    n = len(maps)
    m = len(maps[0])
    visited = [[False] * m for _ in range(n)]

    queue = deque()
    queue.append(0, 0)

    while queue:
        x, y = queue.popleft()

        for i in range(4):

            nx = x + dx[i]
            ny = y + dy[i]

            if 0 <= nx < n and 0 <= ny < m:
                if visited[nx][ny] == False and maps[nx][ny] == 1:
                    maps[nx][ny] = maps[x][y] + 1
                    visited[nx][ny] = True
                    queue.append([nx, ny])
    answer = maps[n-1][m-1]
    if answer == 1:
        answer = -1
    return answer
